### ISSUE TITLE
PlayOnLinux + Wine 1.6.2 (x86) + mgs2-v-s-fix

### ISSUE THREAD

**Issue by inukaze** [VFansss/mgs2-v-s-fix#77]
Hi there, right now i am trying to use your software. i am under GNU/Linux using PlayOnLinux + Wine 1.6.2 of 32 Bits + dotnet45 + your program. [Use dotnet is very awful under wine]

i had an issue when i click settings : 
[![MGS2-S-Wine162.png](https://i.postimg.cc/nhyHmV5j/MGS2-S-Wine162.png)](https://postimg.cc/SnGFF4mk)

I am using this version of wine, because is one of the lastest version with this game works. i know from wine 1.8.6 to the lastest one, the game don't work, when you try to start, just got a black screen and nothing more. never start the intro secuence or something.

This is my GPU Information:
[![Nvidia620.png](https://i.postimg.cc/Lsjm3KKy/Nvidia620.png)](https://postimg.cc/B8QWJzvD)

Basically:
Manufacter : Nvidia
Model : GT 620
Driver Version : 304.137
Operating System : GNU/Linux
Architecture : 64 Bits

**Comment by VFansss** [comment]
Hi!

I'm sorry for the very late reply, but these days are a nightmare for me ☠️ 

Before all: thanks for trying the fix on Linux!

I've tried several times to make the game/fix working under Linux ([I also tried to get the community involved, with little success I have to admit](https://old.reddit.com/r/linux_gaming/comments/9h4lhj/help_me_testing_my_metal_gear_solid_2_fix_on_linux/)) but I'm not good at all with gaming on Tux (my hardware isn't much compatible with it, and I'm not experienced at all) so I had a lot of trouble understanding if my failure was due to myself, the hardware or the software.

So help for more experienced users are crucial.

I've [already encountered the issue you posted in the past](https://old.reddit.com/r/linux_gaming/comments/9h4lhj/help_me_testing_my_metal_gear_solid_2_fix_on_linux/) and, albeit the user didn't replied to me, I've tried to fix the situation.

The problem arise because the fix can't retrieve a list of VGA installed on the system, and because the game executable need to be hardcoded with the VGA name (exactly...the **name** of the VGA), the Fix doesn't know what to do and hang up.

In the past (almost 2 years ago, I think) I've inserted some more methods to retrieve it even under Linux environment, but now I understand that is not working at all, or it fail on your particular configuration. (I bet for the former).

> Use dotnet is very awful under wine

Oh god, you are right 😅 

Well, that is a minor issue for now 😛 

> I am using this version of wine, because is one of the lastest version with this game works. i know from wine 1.8.6 to the lastest one, the game don't work, when you try to start, just got a black screen and nothing more. never start the intro secuence or something.

The game tend to crash into a black screen for a lot of different situations even on windows, because it need a LOT of different patching just to reach the main menu.

Let's focus on making the V's Fix working correctly under Wine, and every issue probably with resolve itself...

> My unique reason for try this, is during videos and cutscenes, i got a screen flickering.

So the game is working or not, without fix? I don't understand...

Let's try this solution before: enable the [debug mode](https://github.com/VFansss/mgs2-v-s-fix/wiki/Troubleshooting-&-Debug-mode#debug-mode) and send me the related log.

Also, I've implemented a fallback method for manually write the VGA name inside the fix:

* Open Configuration_file.ini you can find inside the game 'bin' game folder
* Go to:
`GraphicAdapterName=`
* Write manually your VGA Name. Probably it should like this:
`GraphicAdapterName=GeForce GT 620`
* Save the .ini
* Launch the fix

If is not working, send me the debug file after you have done this also!

**Comment by inukaze** [comment]
Hi there thanks for the answers. well i make a lot of test under GNU/Linux, but all my things are in spanish, for example right now i am test and find solutions for "Wine Bug 9127" in this "[File](https://github.com/inukaze/maestro/blob/master/md/Lista/Videojuegos/Wine-Error-MPEG1.md)"

Metal Gear Solid 2 without your patch works from Wine 1.3.30 until Wine 1.7.54 but need some workaround before you can start the game, like install : Official MGS Patch 2.0, dsound, directmusic, quartz. [From winetricks or playonlinux -> configure -> virtual drive -> tab "Components"] 

If someone want to try your patch should set minimal version of wine on 1.6.2 install the workaround : dotnet40 + dotnet45 [From winetricks or playonlinux -> configure -> virtual drive -> tab "Components"]

"The problem arise because the fix can't retrieve a list of VGA installed on the system, and because the game executable need to be hardcoded with the VGA name (exactly...the name of the VGA), the Fix doesn't know what to do and hang up."

Ok, which values you need to extract exactly?. Because with wine versions happens something very curious, depend of wine version, the "String" of "GPU" is variant. for example on my case : 

Wine 1.2  => Nvidia GeForce 6200
Wine 1.3.30 => Nvidia GeForce 8400
Wine 1.8.6 => Nvidia GeForce 8600

With exactly the same GPU. i think the best way if you can extract information directly from gpu like using "lspci"

For example if you want extract manufacturer from terminal you can use:
`$ lspci -v | grep "VGA" | cut -d " " -f05`

Output : 
`NVIDIA`

For try to extract the model this can be complicated because the variaty of long of names on my case :
```
$ lspci -v | grep "VGA" | cut -d " " -f08-12
$ lspci -v | grep "VGA" -A 12 | head -n1 | awk '{print $8,$9,$10,$11,$12}'
```

Output : 
`[GeForce GT 620] (rev a1)`

But that result can be very variant and useless on some case, i prefer extract "id" for example
`$ lspci -nv | grep "VGA" | cut -d " " -f03`

Output : 
`10de:0f01`

Ok here "**10de**" is "**NVIDIA**" and "**0f01**" is my "**GeForce GT 620 (rev a1)**" you can compare against the information available on the website -> [**PCI IDs**](https://pci-ids.ucw.cz) you can find easely my gpu "[**Here**](https://pci-ids.ucw.cz/read/PC/10de/0f01)"

I think this method is more "**Certero**" (spanish word) [It means that it is more precise, reasonable, informative, specific and according to the truth] for you need to complete more your "Fix Software"

> So the game is working or not, without fix? I don't understand...

well i never test your fix, because i never reach the settings. without your software fix, the game can be started, first if you solve somethings like : sound, music and video reproduction. just work with MGS v2.0 + Wine Fix -> dsound, directmusic, quartz . [this last is for video reproduction]

**Comment by VFansss** [comment]
> 
> Metal Gear Solid 2 without your patch works from Wine 1.3.30 until Wine 1.7.54 but need some workaround before you can start the game, like install : Official MGS Patch 2.0, dsound, directmusic, quartz. [From winetricks or playonlinux -> configure -> virtual drive -> tab "Components"]

This is a good information, and I will keep it in mind...

> If someone want to try your patch should set minimal version of wine on 1.6.2 install the workaround : dotnet40 + dotnet45 [From winetricks or playonlinux -> configure -> virtual drive -> tab "Components"]
> 

That is a good information too: I had a lot of issues with missing libraries when I've tried myself the fix under Wine.

> "(The problem arise because the fix can't retrieve a list of VGA installed on the system, and because the game executable need to be hardcoded with the VGA name (exactly...the name of the VGA), the Fix doesn't know what to do and hang up.")
> 
> Ok, which values you need to extract exactly?. Because with wine versions happens something very curious, depend of wine version, the "String" of "GPU" is variant. 

Uhm...than probably it mean that the right answer to the question is "the one that WINE show up".


> 
> For try to extract the model this can be complicated because the variaty of long of names on my case :
> 
> ```
> $ lspci -v | grep "VGA" | cut -d " " -f08-12
> $ lspci -v | grep "VGA" -A 12 | head -n1 | awk '{print $8,$9,$10,$11,$12}'
> ```
> 
> Output :
> `[GeForce GT 620] (rev a1)`
> 
> But that result can be very variant and useless on some case, i prefer extract "id" for example
> `$ lspci -nv | grep "VGA" | cut -d " " -f03`
> 
> Output :
> `10de:0f01`
> 
> Ok here "**10de**" is "**NVIDIA**" and "**0f01**" is my "**GeForce GT 620 (rev a1)**" you can compare against the information available on the website -> [**PCI IDs**](https://pci-ids.ucw.cz) you can find easely my gpu "[**Here**](https://pci-ids.ucw.cz/read/PC/10de/0f01)" or if for you is better just explore the "[**Vendor Section -> Nvidia**](https://pci-ids.ucw.cz/read/PC/10de)" and if you use the "find" and search "0f01" you can reach to the value you need.
> 
> I think this method is more "**Certero**" (spanish word) [It means that it is more precise, reasonable, informative, specific and according to the truth] for you need to complete more your "Fix Software"

Unfortunately I need the more precise string possible about the VGA, but now I have something to work on.

> well i never test your fix, because i never reach the settings. 

Yup, but it would be fantastic to try the fix I've posted before (write manually the VGA name inside the .ini) to overcome at least that issue, and see if something else show up...

If that is the only issue that the fix has on Linux, I could spend some time adjusting it, but if it has tons of issues well...is a totally different story...

> without your software fix, the game can be started, first if you solve somethings like : sound, music and video reproduction. just work with MGS v2.0 + Wine Fix -> dsound, directmusic, quartz . [this last is for video reproduction]

Well, the Fix overcome the burden to spent a lot of time finding different fixes, put them in place, configure them, and it include nifty things like the gamepad layout setupper and other extras...

Also, a question: are you using a gamepad? Xinput gamepads work good on WINE?

**Comment by NerosTie** [comment]
With Lutris + WIN32 ARCH + Proton 4.2/Wine 4.9 + dotnet45 + steps [here](https://appdb.winehq.org/objectManager.php?sClass=version&iId=7455) ("How to make it works" below).

Well, it works... with some issues: 
with Proton, rumbles don't work but they work with Wine
with Proton, the intro works but not with Wine
there is a crash related to dsound when your are spotted by an enemy, set "Sound Effect Quality" to medium, it solves the issue.

But there is ONE thing which doesn't work: widescreen. Whatever I try, it's stretched. I think it's the patch from nemesis2000 itself. Normally, with wine and those patches, you need to set "d3d8" as native in winecfg, but there is an error then it crashes.

**Comment by VFansss** [comment]
Sorry for the late reply!

I've always been interested in gaming on Linux, but my current laptop doesn't truly work good enough with any distribution, so my knowledge is very low.

Through time, other people has reported me various tactics to make the game/fix work on Linux, but you are the one that has ever reported something that resemble a success.

I will try to support Linux as far as I can, and I have few questions:

- Can you provide some screenshot of the Fix UI? Is working good, or have some glitches?
- The gamepad is working normally? (aside the vibration things you mentioned). What gamepad are you using? Can you please look at [THIS ISSUE](https://github.com/VFansss/mgs2-v-s-fix/issues/29) and tell me if there's been any advancement?
- Is the game have the [green projector bug](https://github.com/VFansss/mgs2-v-s-fix/issues/26) mentioned here?

> But there is ONE thing which doesn't work: widescreen. Whatever I try, it's stretched. I think it's the patch from nemesis2000 itself. Normally, with wine and those patches, you need to set "d3d8" as native in winecfg, but there is an error then it crashes.

In the last 3-4 version of the fix, the widescreen fix is delivered through [Ultimate ASI Loader](https://github.com/ThirteenAG/Ultimate-ASI-Loader): when you enable the "widescreen" from the V's Fix, both Asi loader and WS fix are deployed. 

The first is injected through 'msacm32.dll', and the second through 'dsound_x64.dll'.

Maybe there is some incompatibility with it, and you could investigate.

You could also disable Widescreen fix and AA from the fix, and trying to apply the original widescreen fix from nemesis. You can download it [HERE](https://mega.nz/#!TJB3WYRZ!3CVlz5qShLL7GSRglYuAY2SfPykCoPRE87kwIE6dAQQ)

Please tell us informed: if we reach something truly good, we could make an entire wiki page with tutorial, and for sure I will take in consideration the idea to make next fixes more cross-platform possible!

**Comment by LucesEnBabel** [comment]
Hello.

I'm testing in a clean Proton prefix and I'm getting this error

![image](https://user-images.githubusercontent.com/25283175/60935881-51591780-a299-11e9-98c3-aff71dc45c5b.png)

Is there a list of winetricks packages I need to install in order to make this work?

**Comment by inukaze** [comment]
@LucesEnBabel : Read the first post i mean the game just work in a limited range of wine versions, and Proton is a more newer and modified Wine Version + Dxvk + ¿eSync?. 

Well i say : 

> Metal Gear Solid 2 without your patch works from Wine 1.3.30 until Wine 1.7.54 but need some workaround before you can start the game, like install : Official MGS Patch 2.0, dsound, directmusic, quartz. [From winetricks or playonlinux -> configure -> virtual drive -> tab "Components"]
> 
> If someone want to try your patch should set minimal version of wine on 1.6.2 install the workaround : dotnet40 + dotnet45 [From winetricks or playonlinux -> configure -> virtual drive -> tab "Components"]

Proton the first version was 3.2.0 / Wine 3.2.0 out of the range of wine can start the MGS2, because of that i tell in at the begun that. because i had test wine versions from 0.9.8 to Wine 4.12.1 and **just works with workarounds from Wine 1.3.30 until Wine 1.7.54**.

Proton is a much more newer version of modify wine from the brach 3.X of wine, the old version of Steam Wine was 1.8.5G and just came with some games like : 
1. Septerra Core
2. Enclave
3. Gorky 17
4. X-Blades
5. Realms Of The Haunting [Use a DOSBox emulator for Windows, instead of native for Linux, this is a very absurd for me, because the game works perfectly with native DOSBox 0.74 official]

**Comment by inukaze** [comment]
@LucesEnBabel : Read the third message i specify is very probable the game just run in a range of wine versions 1.3.30 and 1.7.54 with workarounds to make it works. i had test from wine 0.9.8 to wine 4.12 and out the range the game show that message, but is very strange the cuase normally is "DirectMusic" 

Proton 4.2 is a fork of Wine 4.2 + Dxvk + ¿eSync? [i can't remember the name of the last component]

Well the thing its is very probable don't work
-----------------------------------------------------------------------------------------------------------------------------

@VFansss : 

> Unfortunately I need the more precise string possible about the VGA, but now I have something to work on.

This can be very strange because i really don't know from where wine get the information to put to the system registry. and the most strange is depend wine version i use, with my gpu, i can get gpu different strings, depend on branch i try.

Using "OpenSetup" for "Ragnarok Online" but with anothers game the result can be "Nvidia 5200"
Wine 1.2 => Wine D3D T&L HAL
![Wine1 2](https://user-images.githubusercontent.com/1647332/61975323-d0975c80-afb6-11e9-8b5e-74adb8109b17.png)


Wine 1.7.50 => [New] Nvidia GeForce 6800
![Wine1 7 50](https://user-images.githubusercontent.com/1647332/61975355-e3aa2c80-afb6-11e9-977d-6f95682c25cc.png)

**Comment by sianhulo** [comment]
For those having problems geting their card detected, installing vcrun2015 with winetricks fixes that issue, but I still get UAC errors when trying to save. 
And when I try to launch it, I get the same error as @LucesEnBabel when I try to launch.
[MGS2_VFix_debug.txt](https://github.com/VFansss/mgs2-v-s-fix/files/3465683/MGS2_VFix_debug.txt) Here's the log I get.

You asked for screenshots of the ui so here you have one 
![explorer exe 2019-08-04 21:22:25_3](https://user-images.githubusercontent.com/9102033/62434560-b4a66000-b706-11e9-95a0-c3256de28a15.png) the font is aliased but can be seen clearly.

I also decided to try this project https://github.com/Joshua-Ashton/d9vk as it grabs direct 9 and turns it into vulkan, to do this you also have to modify wine so it uses the native d3d8.dll file (by default wine will only use its builtin libraries) created by VFansss' project. Unfortunately, I didn't get far with it.

I get the following prompts after trying to start the game 
![explorer exe 2019-08-04 22:11:44_0](https://user-images.githubusercontent.com/9102033/62434894-419de900-b708-11e9-89f2-dbd7b11e4fe9.png)
![explorer exe 2019-08-04 22:11:51_8](https://user-images.githubusercontent.com/9102033/62434911-51b5c880-b708-11e9-83b2-cdf27e227215.png)
![explorer exe 2019-08-04 22:12:03_3](https://user-images.githubusercontent.com/9102033/62434963-7c078600-b708-11e9-9333-5f35eb4ad102.png)

Now, when running under d9vk, which would only happen when actually trying to launch the game, what we get is that the name of the device changes, which is at least triggering one of the problems/warnings. 
[MGS2_VFix_debug.txt](https://github.com/VFansss/mgs2-v-s-fix/files/3465734/MGS2_VFix_debug.txt)

**Comment by sianhulo** [comment]
As some people might be interested, I had success running the game with wine 4.15 and without applying @VFansss fix, so it might be useful for debugging. I followed the How to on this page https://appdb.winehq.org/objectManager.php?sClass=version&iId=7455 however, I also disabled the "winegstreamer" library through winecfg as otherwise videos won't play.

The game can be finished, but there are a few issues:
*There is a non-zero chance the game will get stuck in an area transition, I think it only happened 4 times to me on a playthrough, you have to close the game because it'll just remain stuck.
*the projectors on the tanker only "project" a green texture.
*you can't see the thumbnails of the digital camera (the photos are still taken, and the camera's "viewport" works normally, you can't just see the thumbnails of taken photos).

After installing dotnet45 through winetricks I was able to install your patch, but it has a few problems under linux:
*it moves the save files but the game still looks for them in the normal folder (maybe the library you use to inject this code isn't being loaded, but I don't know which library it is).
*Not specifically linux, but this overwrote my controller settings (which I had backed up beforehand) to the default ones which are wrong for my controller, so maybe consider not extracting filepadana.ini, padanas.ini, padbtn.ini, padbtns.ini if they already exist.

**Comment by homeisfar** [comment]
I've had success with Linux but this generally not how I'd like to play the game for a couple important reasons. To start with, this is the system & what worked:

Fedora 33, Wine Staging 5.20. I did not use anything but Wine directly.

1) Create a wine 32 bit prefix. Set it as WindowsXP under winecfg.
2) Mount disk image with CDEmu. Install into the wine32 prefix.
3) Use Winetricks to install dsound, directmusic, and quartz DLLs. One of the downloads didn't succeed (sha256sum mismatch) and I had to manually download that file after finding the original link in the winetricks src code.
4) Install DXVK.

From this point the game worked. For some reason mounting through Gnome's default mount tool makes the game think the disk image is unavailable, but using CDEmu resolves that issue, and with it there are no complaints about a missing disc. Most models were invisible but tweaking the graphics lower (adjusting the main sliders 7 units from the left) and adjusting the game resolution brought them back.

5) From here I was able to apply V's fix (once again, running it through the same wine32 prefix).

The important thing when trying to launch the game from MGS2Setup.exe is that your pwd has to be the bin directory. i.e. you must cd to the same directory as the .exe, otherwise it will complain about not finding mgs2.ini.

So now the game works and I'm able to adjust the resolution and all the other good stuff. Generally, it runs fast and smooth. But unfortunately there are two issues: 1) all the sound effects in-game are slightly delayed (consistently like 300ms late). And 2) at least on the controller I tried using, the L2 and R2 buttons did not work. I spent a little time debugging the former without success and I didn't bother with the latter.

**Comment by vosian** [comment]
Well, regarding the controller, my memory is fuzzy but there should be a file with button mappings, unfortunately I think they were hexcode, but if I recall, it just went from 0 to number of buttons. I'd check with wine's joystick tester thing to make sure the buttons are detected, and if they are, play with the text file with the mappings
